toffy/normalize.py

```diff
@@ -0,0 +1,238 @@
+        combined = pulse_heights.merge(channel_counts, 'outer', on=['fovs', 'masses'])
+
+        if len(combined) != len(pulse_heights):
+            raise ValueError("Pulse heights and channel counts must be generated for the same "
+                             "mass ranges and fovs. However, the data the following does do not "
+                             "exactly match: {}".format(dir))
+
+        # add directory label and add to list
+        combined['directory'] = dir
+        all_dirs.append(combined)
+
+    # combine data from each dir together
+    all_data = pd.concat(all_dirs)
+    all_data.reset_index()
+
+    # create normalized counts column
+    subset = all_data[['channel_counts', 'masses']]
+    all_data['norm_channel_counts'] = subset.groupby('masses').transform(lambda x: (x / x.max()))
+
+    return all_data
+
+
+def normalize_image_data(data_dir, output_dir, fovs, pulse_heights, panel_info,
+                         norm_func_path):
+    """Normalizes image data based on median pulse height from the run and a tuning curve
+
+    Args:
+        data_dir (str): directory with the image data
+        output_dir (str): directory where the normalized images will be saved
+        fovs (list or None): which fovs to include in normalization. If None, uses all fovs
+        pulse_heights: file containing pulse heights per mass per fov
+        panel_info_path: file containing mapping between channels and masses
+        norm_func_path: file containing the saved weights for the normalization function
+    """
+
+    # get FOVs to loop over
+    if fovs is None:
+        fovs = io_utils.list_folders(data_dir)
+
+    # load calibration function
+    with open(norm_func_path, 'r') as cf:
+        norm_json = json.load(cf)
+
+    norm_weights, norm_name = norm_json['weights'], norm_json['name']
+
+    channels = panel_info['targets'].values
+
+    # instantiate function which translates pulse height to a normalization constant
+    norm_func = create_prediction_function(norm_name, norm_weights)
+
+    for fov in fovs:
+        output_fov_dir = os.path.join(output_dir, fov)
+        os.makedirs(output_fov_dir)
+
+        # get images and pulse heights for current fov
+        images = load_utils.load_imgs_from_tree(data_dir, fovs=[fov], channels=channels,
+                                                dtype='float32')
+        fov_pulse_heights = pulse_heights.loc[pulse_heights['fov'] == fov, :]
+
+        # fit a function to model pulse height as a function of mass
+        mph_weights = fit_calibration_curve(x_vals=fov_pulse_heights['masses'].values,
+                                            y_vals=fov_pulse_heights['mphs'].values,
+                                            obj_func='poly_2')
+
+        # predict mph for each mass in the panel
+        mph_func = create_prediction_function(name='poly_2', weights=mph_weights)
+        mph_vals = mph_func(panel_info['masses'].values)
```

Comment:
Sure.  I guess I'm saying that `mph_func` can be retrieved **once**, outside the loop, directly from `create_objective_function` and be called via:

```
mph_vals = mph_func(panel_info['masses'].values, *mph_weights)
```

This creates only one, reused function object, while the current implementation creates a function object per fov.

Reply:
Got it. I don't think there's any real efficiency gain, given how small these functions are. Do you think it's syntactically better to create a single function and update the weights within each iteration of the loop, rather than create a function with the weights within the loop? To me they look similarly interpretable.

Reply:
since this is qc, and run on a 'per-run' basis (~20 fovs), this doesn't matter much.  In terms of best practices/legibility, I think opting out of a function factory when possible is best, but yeah doesn't matter much either way.